src/hypothesis/internal/conjecture/minimizer.py

```diff
@@ -116,79 +94,130 @@ def rotate_suffixes(self):
+                # We first see if replacing the byte with zero and the rest
+                # with 255 is enough to trigger the condition. If this
+                # succeeds we've successfully zeroed the byte here and the
+                # rest of this search isn't useful.
+                not suitable(0) and
+
+                # We now check if the lexicographic predecessor (where this
+                # element is reduced by 1 and all subsequent elements are
+                # raised to 255) is valid here. If it's not then there's
+                # no point in trying the search and we can break out early.
+                suitable(c - 1)
+            ):
+                # We now do a binary search to find a small value
+                # where the large suffix works. Again, the property is not
+                # necessarily monotonic, so this doesn't actually guarantee
+                # the smallest value.
+                @binsearch(0, self.current[i])
+                def _(m):
+                    # We have to manually check the end point because we
+                    # already incorporated this.
+                    if m == self.current[i]:
+                        return True
+                    return suitable(m)
+
+            # We may have replaced the suffix with a block of 255s. See if
+            # we can replace it with our old suffix, which might be a lot
+            # more friendly.
+            self.incorporate(self.current[:i + 1] + original_suffix)
 
     def run(self):
         if not any(self.current):
             return
         if len(self.current) == 1:
-            for c in range(self.current[0]):
+            for c in hrange(self.current[0]):
                 if self.incorporate(hbytes([c])):
                     break
             return
+
+        # Initial checks as to whether the two smallest possible buffers of
+        # this length can work. If so there's nothing to do here.
         if self.incorporate(hbytes(self.size)):
             return
-        if self.incorporate(hbytes([0] * (self.size - 1) + [1])):
-            return
+
+        if not self.cautious or self.current[-1] > 0:
+            if self.incorporate(hbytes([0] * (self.size - 1) + [1])):
+                return
+
+        # Perform a binary search to try to replace a long initial segment with
+        # zero bytes.
+        # Note that because this property isn't monotonic this will not always
+        # find the longest subsequence we can replace with zero, only some
+        # subsequence.
+
+        # Replacing the first nonzero bytes with zero does *not* work
+        nonzero = len(self.current)
+
+        # Replacing the first canzero bytes with zero *does* work.
+        canzero = 0
+        while self.current[canzero] == 0:
+            canzero += 1
+
+        @binsearch(canzero, nonzero)
+        def zero_prefix(mid):
+            return self.incorporate(
+                self.current[:mid] + hbytes(len(self.current) - mid)
```

Comment:
You're entirely right. The shift_right logic below is also wrong in the same way!

I clearly wasn't fully possessed of brain when I wrote this code.